File: CodingSamples/DotNet/Fundamentals/Inheritance/InterfaceTypeTest/Banking/SavingsAccount.cs
namespace Banking;

//multiple inheritance
class SavingsAccount : Account, IProfitable
{
    //an identifier declared with const keyword is replaced
    //by its initialier value wher eever it appears in code
    const decimal MinBal = 5000;

    internal SavingsAccount()
    {
        Balance = MinBal;
    }

    public override void Deposit(decimal amount)
    {
        if(Balance >= 50000)
            amount -= 100;
        Balance += amount;
    }

    //a method declared with sealed keyword cannot be
    //further overridden in the derived class
    public sealed override void Withdraw(decimal amount)
    {
        if(Id < 0)
            throw new NotSupportedException("Account is locked.");
        if(Balance - amount < MinBal)
            throw new InsufficientFundsException();
        Balance -= amount;
    }

    public decimal AddInterest(int months)
    {
        decimal rate = Balance < 5 * MinBal ? 3 : 4;
        decimal interest = Balance * months * rate / 1200;
        Balance += interest;
        return interest;
    }
}

File: CodingSamples/Java/Fundamentals/Inheritance/InterfaceTypeTest/bank/banking/SavingsAccount.java
package banking;

final class SavingsAccount extends Account implements Profitable{
    
    //a field defined with final keyword cannot be reassigned
    static final double MIN_BAL = 5000;

    SavingsAccount() {
        balance = MIN_BAL;
    }

    public void deposit(double amount) {
        balance += amount;
    }

    public void withdraw(double amount) throws InsufficientFundsException {
        if(balance - amount < MIN_BAL)
            throw new InsufficientFundsException();
        balance -= amount;
    }

    public double interest(int months) {
        double rate = balance < 5 * MIN_BAL ? MIN_RATE : MIN_RATE + 0.5;
        return balance * rate * months / 1200;
    }
}
